contents/blog/how-we-improved-feature-flags-resiliency.md

```diff
@@ -0,0 +1,145 @@
+---
+date: 2023-03-28
+title: How we improved feature flag resiliency
+rootPage: /blog
+sidebar: Blog
+showTitle: true
+---
+
+Feature flags as a service is an interesting space to be working in. If your service stops working, not only does it affect your customers, but also your
+customer's customers, since they rely on you to make sure their app works fine.
+
+In contrast to say, the PostHog interface not loading, where the problem is constrained to your customers. It's not great, for sure, but it's better than event ingestion and feature flags going down.
+
+Further, flags can be very sensitive to latency: If it takes 5 seconds for your flags to evaluate, that is you holding up your customers application for 5 seconds. No, you can't wait to load them asynchronously as you need this result to determine what to show. Your business logic depends on this value.
+
+So, over the past quarter, one of our key goals has been to make feature flags fast and resilient: Even if the PostHog interface goes down, none of the SDKs that query flags should go down, nor the pure API interface. Further, for the strict latency flags, we need a way to resolve them in 50ms.
+
+This post is about how we did it, and what we learned along the way.
+
+Before we jump into how we improved things, it's worth listing out constraints that we can use to create solutions that otherwise wouldn't be possible.
+
+
+## Special Problem Constraints
```

Comment:
To help people get to the "meat" of the article, it might be worth it just making this section a bullet point list like
- Flags are deterministic: If I roll out a flag to 30% of people, the person who's inside this window remains inside this window even if the rollout percentage goes up to 40, 80, or 100%. This is because we compute a hash of the user id and the flag key and use it evaluate. 
- Flags are evaluated multiple times in a session: flags are evaluated multiple times on a page. For example, whenever properties change our flag might change, and same for when the user identity changes. We can leverage this behaviour in our solution.
- The Caching Problem
...
etc.

Reply:
Basically I want people to read the "How we made feature flags faster and more reliable" part of the article, and this section prevents that 😅

Reply:
Makes sense, I like it